### DISCUSSION TITLE
Tenant interface for the future of the package

### DISCUSSION THREAD

**Discussion by masterix21** [spatie/laravel-multitenancy#543] [Ideas]
Hi @freekmurze, I have thought about laravel-multitenancy in the last few weeks. 

The model `Tenant` is okay, but it is simplistic when you need to add a bit of complexity.

My idea is that:
- We need an interface `IsTenant`
- We need a trait `HasTenantFeatures`

Starting from that, our start model `Tenant` can change like so:
```php
class Tenant extends Models implements IsTenant
{
    use UsesLandlordConnection;
    use HasFactory;
    use HasTenantFeatures;
}
```

In the `HasTenantFeatures`, we move the methods:
- `makeCurrent() : IsTenant`
- `forget() : IsTenant`
- `current() : ?IsTenant`
- `checkCurrent() : bool`
-  `isCurrent(): bool`
- `forgetCurrent(): ?IsTenant`
- `getDatabaseName(): string`
- `newCollection(array $models = []): TenantCollection`
- `execute(callable $callable) : mixed`
- `callback(callable $callable): \Closure`

We replace all references from Tenant to IsTenant.

I'm sure these changes will lessen the difficulty reported by many users in the last few years.

What do you think about it?

PS: If you say `yes`, I'm ready to improve the code: no stress for you.

References:
- https://github.com/spatie/laravel-multitenancy/discussions/443

**Comment by masterix21** [comment]
For example, using Laravel Jetstream, you could add tenant features to `Team` model like so:

```php
class Team extends \Laravel\Jetstream\Team implements IsTenant
{
    use HasFactory;
    use UsesLandlordConnection;
    use HasTenantFeatures;
}
```

**Comment by freekmurze** [comment]
Great idea, I would be ok with going through with this (and tagging a new major version for this)
